src/inference/AdvancedSMC.jl

```diff
@@ -17,76 +43,106 @@ must be stored in a [`TArray`](@ref) object.
 Usage:
 
 ```julia
-SMC(1000)
+SMC()
 ```
 """
-mutable struct SMC{space, F} <: InferenceAlgorithm
-    n_particles           ::  Int
+mutable struct SMC{space, F} <: ParticleInference
     resampler             ::  F
     resampler_threshold   ::  Float64
 end
-function SMC(n_particles::Int, resampler::F, resampler_threshold::Float64, space::Tuple) where {F}
-    return SMC{space, F}(n_particles, resampler, resampler_threshold)
+
+alg_str(spl::Sampler{SMC}) = "SMC"
+function SMC(
+    resampler::F,
+    resampler_threshold::Float64,
+    space::Tuple
+) where {F}
+    return SMC{space, F}(resampler, resampler_threshold)
 end
-SMC(n) = SMC(n, resample_systematic, 0.5, ())
-SMC(n_particles::Int, ::Tuple{}) = SMC(n_particles)
-function SMC(n_particles::Int, space::Symbol...)
-    SMC(n_particles, resample_systematic, 0.5, space)
+SMC() = SMC(resample_systematic, 0.5, ())
+SMC(::Tuple{}) = SMC()
+function SMC(space::Symbol...)
+    SMC(resample_systematic, 0.5, space)
 end
 
-function Sampler(alg::SMC, s::Selector)
-    info = Dict{Symbol, Any}()
-    info[:logevidence] = []
-    return Sampler(alg, info, s)
+mutable struct SMCState{V<:VarInfo, F<:AbstractFloat} <: AbstractSamplerState
+    vi                   ::   V
+    # The logevidence after aggregating all samples together.
+    average_logevidence  ::   F
+    particles            ::   ParticleContainer
 end
 
-function step(model, spl::Sampler{<:SMC}, vi::VarInfo)
+function SMCState(
+    model::M, 
+) where {
+    M<:Model
+}
+    vi = VarInfo(model)
     particles = ParticleContainer{Trace}(model)
-    vi.num_produce = 0;  # Reset num_produce before new sweep\.
-    set_retained_vns_del_by_spl!(vi, spl)
-    resetlogp!(vi)
 
-    push!(particles, spl.alg.n_particles, spl, vi)
+    return SMCState(vi, 0.0, particles)
+end
 
-    while consume(particles) != Val{:done}
-      ess = effectiveSampleSize(particles)
-      if ess <= spl.alg.resampler_threshold * length(particles)
-        resample!(particles,spl.alg.resampler)
-      end
-    end
+function Sampler(alg::T, model::Model, s::Selector) where T<:SMC
+    dict = Dict{Symbol, Any}()
+    state = SMCState(model)
+    return Sampler(alg, dict, s, state)
+end
 
-    ## pick a particle to be retained.
-    Ws, _ = weights(particles)
-    indx = randcat(Ws)
-    push!(spl.info[:logevidence], particles.logE)
+function sample_init!(
+    ::AbstractRNG, # Note: This function does not use the range argument.
+    model::Turing.Model,
+    spl::Sampler{<:SMC},
+    N::Integer; # Note: This function doesn't use the N argument.
+    kwargs...
+)
+    # Set the parameters to a starting value.
+    initialize_parameters!(spl; kwargs...)
 
-    return particles[indx].vi, true
-end
+    # Update the particle container now that the sampler type
+    # is defined.
+    spl.state.particles = ParticleContainer{Trace{typeof(spl),
```

Comment:
I can't figure out a way to do this in an elegant way, since the types are highly nested. I think for now `typeof` satisfies all use cases without adding a fairly large parameterizion block at the function declaration. I'm open to tips on this one -- maybe I'm missing something here.

Reply:
Fine as is for me. We can try getting rid of the typeof statements later.